go/libraries/doltcore/sqle/setalgebra/interval.go

```diff
@@ -0,0 +1,293 @@
+// of 0 indicates equality, and a value of 1 indicates the value in interval 1 is greater than
+// the value in interval 2.
+func compareIntervals(in1, in2 Interval) (intervalComparison, error) {
+	comp := intervalComparison{}
+	endpoints := [2][2]*IntervalEndpoint{{in1.Start, in1.End}, {in2.Start, in2.End}}
+
+	var err error
+	for i := 0; i < 2; i++ {
+		for j := 0; j < 2; j++ {
+			ep1, ep2 := endpoints[0][i], endpoints[1][j]
+			lt, eq := false, false
+
+			if ep1 == nil && ep2 == nil {
+				// if both points are null they are only equivalent when comparing start points
+				// or end points, and they are less when comparing a start point to an end point
+				// but greater when comparing an end point to a start point
+				if i == j {
+					eq = true
+				} else {
+					lt = i < j
+				}
+			} else if ep1 == nil {
+				// if an intervalEndpoint is nil in the first point it will be less than the
+				// second point if it is a nil start point. In all other cases it is greater,
+				// and it can never be equal to a non nil intervalEndpoint
+				lt = i == 0
+			} else if ep2 == nil {
+				// if an intervalEndpoint is nil in the second point then the first point will be less
+				// if it is a nil end point. In all other cases it is greater greater, and it can never
+				// be equal to a non nil intervalEndpoint
+				lt = j == 1
+			} else {
+				// compare 2 valid intervalEndpoints
+				eq = ep1.Val.Equals(ep2.Val)
+
+				if eq {
+					if !ep1.Inclusive && !ep2.Inclusive {
+						// If equal, but both intervalEndpoints are open, they are only equal if comparing two
+						// start points or to end points. Otherwise they are not equal.
+						if i != j {
+							eq = false
+							lt = i > j
+						}
+					} else if !ep1.Inclusive {
+						// intervalEndpoints are not equal unless both are open, or both are closed
+						eq = false
+						lt = i == 1
+					} else if !ep2.Inclusive {
+						// intervalEndpoints are not equal unless both are open, or both are closed
+						eq = false
+						lt = j == 0
+					}
+				} else {
+					// both points are non nil and not equal so simply check to see if the first point is less
+					// than the second.
+					lt, err = ep1.Val.Less(in1.nbf, ep2.Val)
+
+					if err != nil {
+						return intervalComparison{}, err
+					}
+				}
+			}
+
+			var res int
+			if lt {
+				res = -1
+			} else if !eq {
+				res = 1
+			}
+
+			resIdx := i*2 + j
+			comp[resIdx] = res
+		}
+	}
+
+	return comp, nil
+}
+
+// simplifyInterval will return:
+//  * a UniversalSet for an equivalent interval defined as: negative infinity < X < positive infinity
+//  * a FiniteSet with a single value N for an equivalent interval defined as: N <= X <= N
+//  * EmptySet for an interval defined as: N < X < N
+//  * EmptySet for an interval where end < start
+//  * an unchanged interval will be returned for all other conditions
+func simplifyInterval(in Interval) (Set, error) {
```

Comment:
I would consider removing this, not sure it's pulling its weight. Seems like a premature optimization.